compiler/rustc_codegen_ssa/src/sir.rs

```diff
@@ -341,102 +289,292 @@ impl SirFuncCx<'tcx> {
-            mir::Rvalue::Cast(mir::CastKind::Misc, op, ty) => self.lower_cast_misc(op, ty),
-            _ => ykpack::Rvalue::Unimplemented(with_no_trimmed_paths(|| {
+            mir::Rvalue::Cast(mir::CastKind::Misc, op, ty) => self.lower_cast_misc(bx, bb, op, ty),
+            _ => ykpack::IPlace::Unimplemented(with_no_trimmed_paths(|| {
                 format!("unimplemented rvalue: {:?}", rvalue)
             })),
         }
     }
 
-    pub fn lower_place(&self, place: &mir::Place<'_>) -> ykpack::Place {
-        ykpack::Place {
-            local: self.lower_local(place.local),
-            // FIXME projections not yet implemented.
-            projection: place.projection.iter().map(|p| self.lower_projection(&p)).collect(),
+    fn monomorphize<T>(&self, value: &T) -> T
+    where
+        T: TypeFoldable<'tcx> + Copy,
+    {
+        if let Some(substs) = self.instance.substs_for_mir_body() {
+            self.tcx.subst_and_normalize_erasing_regions(substs, ty::ParamEnv::reveal_all(), value)
+        } else {
+            self.tcx.normalize_erasing_regions(ty::ParamEnv::reveal_all(), *value)
         }
     }
 
-    pub fn lower_projection(&self, pe: &mir::PlaceElem<'_>) -> ykpack::Projection {
-        match pe {
-            mir::ProjectionElem::Field(field, ..) => ykpack::Projection::Field(field.as_u32()),
-            mir::ProjectionElem::Deref => ykpack::Projection::Deref,
-            mir::ProjectionElem::Index(local) => {
-                ykpack::Projection::Index(self.lower_local(*local))
-            }
-            _ => ykpack::Projection::Unimplemented(format!("{:?}", pe)),
-        }
+    /// Wrapper for bx.layout_of() which ensures the type is first monomorphised.
+    fn layout_of<Bx: BuilderMethods<'a, 'tcx>>(&self, bx: &Bx, t: Ty<'tcx>) -> TyAndLayout<'tcx> {
+        bx.layout_of(self.monomorphize(&t))
     }
 
-    pub fn lower_local(&self, local: mir::Local) -> ykpack::Local {
-        // For the lowering of `Local`s we currently assume a 1:1 mapping from MIR to SIR. If this
-        // mapping turns out to be impossible or impractial, this is the place to change it.
-        ykpack::Local(local.as_u32())
+    /// Apply an offset to an IPlace.
+    fn offset_iplace<Bx: BuilderMethods<'a, 'tcx>>(
+        &mut self,
+        bx: &Bx,
+        ip: &mut ykpack::IPlace,
+        add: ykpack::OffT,
+        mirty: Ty<'tcx>,
+    ) {
+        match ip {
+            ykpack::IPlace::Val { off, ty, .. } => {
+                *off += add;
+                *ty = self.lower_ty_and_layout(bx, &self.layout_of(bx, mirty));
+            }
+            ykpack::IPlace::Indirect { off, ty, .. } => {
+                *off += add;
+                *ty = self.lower_ty_and_layout(bx, &self.layout_of(bx, mirty));
+            }
+            ykpack::IPlace::Const { .. } => {
+                // Oddly enough this happens. It seems to be only ever with an offset of 0 and we
```

Comment:
Comment doesn't parse.

Reply:
Actually I disagree that the type doesn't change. I'll fix that too.

Reply:
Also fixed in 1bd0a0fb698821f6c776591f9f821ab3bbc84811

Sorry the fixes are mixed.